### PR TITLE
Support generation of paged response classes

### PR DESCRIPTION
**Pull request by michaelbausor**
This is a WIP that can hopefully be simplified further. It adds support for generated page response types, with methods to iterate elements as resource names.

### CODE REVIEW

**Comment by michaelbausor** [comment]
See also https://github.com/googleapis/gax-java/pull/117

**Comment by garrettjonesgoogle** [comment]
I'll let @shinfan do the first pass

File: src/main/java/com/google/api/codegen/transformer/PageStreamingTransformer.java [hunk 1]
```diff
@@ -117,4 +118,43 @@ private PageStreamingDescriptorClassView generateDescriptorClass(
...
+  }
+
+  private PageStreamingFactoryClassView generateFactoryClass(MethodTransformerContext context) {
+    SurfaceNamer namer = context.getNamer();
+    ModelTypeTable typeTable = context.getTypeTable();
+    Method method = context.getMethod();
+    PageStreamingConfig pageStreaming = context.getMethodConfig().getPageStreaming();
+    FeatureConfig featureConfig = context.getFeatureConfig();
+
+    PageStreamingFactoryClassView.Builder fact = PageStreamingFactoryClassView.newBuilder();
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/PageStreamingTransformer.java]: s/fact/factory

File: src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java [hunk 2]
```diff
@@ -616,6 +631,7 @@ public String getGetResourceListCallName(Field resourcesField) {
    * resources field, saves it in the given type table, and returns it.
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java]: Update the comments

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpSurfaceNamer.java [hunk 3]
```diff
@@ -85,6 +85,7 @@ public String getFieldGetFunctionName(TypeRef type, Name identifier) {
 
   @Override
   public String getAndSavePagedResponseTypeName(
+      Method method,
       FeatureConfig featureConfig,
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/csharp/CSharpSurfaceNamer.java]: Can we remove this featureConfig since you removed in L98 (assuming it is not used else where)

File: src/main/java/com/google/api/codegen/transformer/PageStreamingTransformer.java [hunk 4]
```diff
@@ -117,4 +118,43 @@ private PageStreamingDescriptorClassView generateDescriptorClass(
...
+
+    Field resourceField = pageStreaming.getResourcesField();
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/PageStreamingTransformer.java]: Move this line up to be together with other init lines.

File: src/main/java/com/google/api/codegen/transformer/PageStreamingTransformer.java [hunk 5]
```diff
@@ -117,4 +118,43 @@ private PageStreamingDescriptorClassView generateDescriptorClass(
...
+    fact.name(namer.getPageStreamingDescriptorConstName(method));
+    fact.requestTypeName(typeTable.getAndSaveNicknameFor(method.getInputType()));
+    fact.responseTypeName(typeTable.getAndSaveNicknameFor(method.getOutputType()));
+    fact.resourceTypeName(typeTable.getAndSaveNicknameForElementType(resourceField.getType()));
+    fact.pagedListResponseTypeName(
+        namer.getAndSavePagedResponseTypeName(
+            method,
+            featureConfig,
+            typeTable,
+            method.getInputType(),
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/PageStreamingTransformer.java]: I think it is safe to assume these types are always the input and output types of the given method. Since you have already passed in the method, can we get rid of the input and output types here?

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java [hunk 6]
```diff
@@ -165,6 +176,82 @@ private StaticLangXApiView generateXApi(SurfaceTransformerContext context) {
...
   }
 
+  private List<StaticLangPagedResponseView> generatePagedResponseWrappers(
+      SurfaceTransformerContext context) {
+    List<StaticLangPagedResponseView> pagedResponseWrappers = new ArrayList<>();
+
+    for (Method method : context.getSupportedMethods()) {
+      MethodTransformerContext methodContext = context.asMethodContext(method);
+      MethodConfig methodConfig = context.getMethodConfig(method);
+      ModelTypeTable typeTable = context.getTypeTable();
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java]: Shall we use a separate type table for each wrapper?

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java [hunk 7]
```diff
@@ -165,6 +176,82 @@ private StaticLangXApiView generateXApi(SurfaceTransformerContext context) {
     return xapiClass.build();
   }
 
+  private List<StaticLangPagedResponseView> generatePagedResponseWrappers(
+      SurfaceTransformerContext context) {
+    List<StaticLangPagedResponseView> pagedResponseWrappers = new ArrayList<>();
+
+    for (Method method : context.getSupportedMethods()) {
+      MethodTransformerContext methodContext = context.asMethodContext(method);
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java]: Move these init lines into the if block

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java [hunk 8]
```diff
@@ -165,6 +176,82 @@ private StaticLangXApiView generateXApi(SurfaceTransformerContext context) {
...
+      MethodConfig methodConfig = context.getMethodConfig(method);
+      ModelTypeTable typeTable = context.getTypeTable();
+
+      if (methodConfig.isPageStreaming()) {
+        StaticLangPagedResponseView.Builder pagedResponseWrapper =
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java]: This building part is quite long. Maybe worth creating a helper method - generatePagedResponseWrapper(MethodTransformerContext context)

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java [hunk 9]
```diff
@@ -165,6 +176,82 @@ private StaticLangXApiView generateXApi(SurfaceTransformerContext context) {
...
+            typeTable.getAndSaveNicknameFor(method.getOutputType()));
+        pagedResponseWrapper.resourceTypeName(
+            typeTable.getAndSaveNicknameForElementType(resourceField.getType()));
+        pagedResponseWrapper.iterateMethods(getIterateMethods(methodContext));
+        pagedResponseWrapper.imports(new ArrayList<ImportTypeView>());
+
+        String outputPath =
+            pathMapper.getOutputPath(context.getInterface(), context.getApiConfig());
+        pagedResponseWrapper.outputPath(
+            outputPath + File.separator + pagedResponseTypeName + ".java");
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java]: If an API is large we may generate lots of such paged response files (Most of them would be empty classes). I suggest we put these files under a sub-folder so  they do not pollute the main files, e.g. outputPath/pagestreaming

**Comment by shinfan** [comment]
The overall approach looks okay to me. Just a few comments.

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java (hunk 9 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java]: Hmmm, yes I think this might be a good idea. Another option is to put them all into the same file, possibly as inner classes.

Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java]: Yea, I think putting everything into a top class would be better. Probably something like PagedResponseWrappers.java

File: src/test/java/com/google/api/codegen/testdata/java_test_library.baseline [hunk 10]
```diff
@@ -668,9 +668,9 @@ public class LibraryServiceTest {
     List<BookName> names = Arrays.asList(namesElement);
     List<ShelfName> shelves = new ArrayList<>();
 
-    PagedListResponse<FindRelatedBooksRequest, FindRelatedBooksResponse, BookName> pagedListResponse = api.findRelatedBooks(names, shelves);
+    FindRelatedBooksPagedResponse pagedListResponse = api.findRelatedBooks(names, shelves);
 
-    List<BookName> resources = Lists.newArrayList(pagedListResponse.iterateAllElements());
+    List<String> resources = Lists.newArrayList(pagedListResponse.iterateAllElements());
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/java_test_library.baseline]: What happened to the `BookName` type here?

File: src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline [hunk 11]
```diff
@@ -930,7 +976,7 @@ public class LibraryServiceSettings extends ServiceApiSettings {
     /**
      * Returns the builder for the settings used for calls to listShelves.
      */
-    public PageStreamingCallSettings.Builder<ListShelvesRequest, ListShelvesResponse, Shelf> listShelvesSettings() {
+    public PageStreamingCallSettings.Builder<ListShelvesRequest, ListShelvesResponse, Shelf, ListShelvesPagedResponse> listShelvesSettings() {
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline]: This looks out of date. Maybe generate into api-client-staging and run the java tests to make sure everything is consistent.

File: src/test/java/com/google/api/codegen/testdata/java_test_library.baseline (hunk 10 above)
Comment (michaelbausor) [inline src/test/java/com/google/api/codegen/testdata/java_test_library.baseline]: iterateAllElements() now always iterates over the basic type (strings), and FindRelatedBooksPagedResponse has an additional method iterateAllAsBookName(). I will add a test for this method as well

File: src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline (hunk 11 above)
Comment (michaelbausor) [inline src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline]: Yes this is wrong, I will update and test

File: src/test/java/com/google/api/codegen/testdata/java_page_streaming_response_library.baseline [hunk 12]
```diff
@@ -45,9 +60,25 @@ public class FindRelatedBooksPagedResponse extends PagedListResponse<FindRelated
...
+import com.google.example.library.v1.ListBooksRequest;
+import com.google.example.library.v1.ListBooksResponse;
 
 @javax.annotation.Generated("by GAPIC")
-public class ListBooksPagedResponse extends PagedListResponse<ListBooksRequest, ListBooksResponse, Book> {
+public class ListBooksPagedResponse extends PagedListResponseImpl<ListBooksRequest, ListBooksResponse, Book> {
+
+  public ListBooksPagedResponse(
+      ListBooksRequest request,
+      FutureCallable<ListBooksRequest, ListBooksResponse> callable,
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/java_page_streaming_response_library.baseline]: I would prefer to have this take a UnaryApiCallable (and consequently, PagedListResponseImpl also). You can easily wrap in the place where the FutureCallable is created by doing UnaryApiCallable.create(futureCallable).

Comment (michaelbausor) [inline src/test/java/com/google/api/codegen/testdata/java_page_streaming_response_library.baseline]: Done

**Comment by shinfan** [comment]
@michaelbausor Have you experimented with putting all the pagedList classes under a top class?

**Comment by michaelbausor** [comment]
Not yet - I will try this now.

**Comment by michaelbausor** [comment]
@shinfan Have made that change. The only downside is that now the class names are more verbose, e.g. PagedResponseWrapper.ListTopicSubscriptionsPagedResponse

File: src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline [hunk 13]
```diff
@@ -139,24 +139,24 @@ public class LibraryServiceSettings extends ServiceApiSettings {
 
   private final SimpleCallSettings<CreateShelfRequest, Shelf> createShelfSettings;
   private final SimpleCallSettings<GetShelfRequest, Shelf> getShelfSettings;
-  private final PageStreamingCallSettings<ListShelvesRequest, ListShelvesResponse, ListShelvesPagedResponse> listShelvesSettings;
+  private final PageStreamingCallSettings<ListShelvesRequest, ListShelvesResponse, PagedResponseWrappers.ListShelvesPagedResponse> listShelvesSettings;
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline]: Maybe we could do a static import so that this doesn't get unwieldy? It might require some fancy typing work.

File: src/test/java/com/google/api/codegen/testdata/java_page_streaming_response_library.baseline [hunk 14]
```diff
@@ -18,142 +18,75 @@ import com.google.api.gax.grpc.CallContext;
...
+        ListStringsRequest request,
+        UnaryApiCallable<ListStringsRequest, ListStringsResponse> callable,
+        PageStreamingDescriptor<ListStringsRequest, ListStringsResponse, String> pageDescriptor,
+        CallContext context) {
+      super(request, callable, pageDescriptor, context);
+    }
 
 
+  }
+  public static class FindRelatedBooksPagedResponse extends PagedListResponseImpl<FindRelatedBooksRequest, FindRelatedBooksResponse, String> {
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/java_page_streaming_response_library.baseline]: add a blank line between classes

File: src/main/resources/com/google/api/codegen/java/page_streaming_response.snip [hunk 15]
```diff
@@ -1,31 +1,35 @@
...
-      }
+  public class {@pagedResponseClasses.name} {
   
-      {@iterateMethods(pagedResponseClass)}
-    }
-  @end
+    @join pagedResponseClass : pagedResponseClasses.pagedResponseWrapperList
+      @let requestT = {@pagedResponseClass.requestTypeName}, \
+          responseT = {@pagedResponseClass.responseTypeName}, \
+          resourceT = {@pagedResponseClass.resourceTypeName}
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/page_streaming_response.snip]: don't use let variables in snippets.

File: src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline (hunk 13 above)
Comment (shinfan) [inline src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline]: +1. The name is definitely too long. The other plan is to have all the classes under a subfolder.

Comment (michaelbausor) [inline src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline]: Added support for static imports

File: src/test/java/com/google/api/codegen/testdata/java_page_streaming_response_library.baseline (hunk 14 above)
Comment (michaelbausor) [inline src/test/java/com/google/api/codegen/testdata/java_page_streaming_response_library.baseline]: Done

File: src/main/resources/com/google/api/codegen/java/page_streaming_response.snip (hunk 15 above)
Comment (michaelbausor) [inline src/main/resources/com/google/api/codegen/java/page_streaming_response.snip]: Done

File: src/main/resources/com/google/api/codegen/java/common.snip [hunk 16]
```diff
@@ -25,3 +25,16 @@
     {@arg}
   @end
 @end
+
+@snippet importList(imports)
+  @join import : imports
+    @switch import.type.toString
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/common.snip]: you don't need the toString here

File: src/main/java/com/google/api/codegen/util/java/JavaTypeTable.java [hunk 17]
```diff
@@ -81,6 +83,12 @@ public TypeName getTypeName(String fullName) {
   }
 
   @Override
+  public TypeName getTypeNameFromShortName(String shortName) {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/util/java/JavaTypeTable.java]: It is not clear from the name what this does. It should probably be something like 'getTypeNameInImplicitPackage'.

File: src/main/java/com/google/api/codegen/util/go/GoTypeTable.java [hunk 18]
```diff
@@ -44,6 +43,11 @@ public TypeName getTypeName(String fullName) {
   }
 
   @Override
+  public TypeName getTypeNameFromShortName(String shortName) {
+    throw new UnsupportedOperationException("getTypeNameFromShortName not supported by Go");
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/util/go/GoTypeTable.java]: Why can't we support this?

File: src/main/java/com/google/api/codegen/transformer/java/JavaSurfaceNamer.java [hunk 19]
```diff
@@ -126,15 +126,17 @@ public String getGenericAwareResponseTypeName(TypeRef outputType) {
...
     // TODO(michaelbausor) make sure this uses the typeTable correctly
-    return getPagedResponseWrappersClassName()
-        + "."
-        + getAndSavePagedResponseTypeInnerName(method, typeTable, resourceField);
+
+    String fullPackageWrapperName = typeTable.getFullNameFor(getPagedResponseWrappersClassName());
+    return typeTable.getAndSaveNicknameForStaticInnerClass(
+        fullPackageWrapperName
+            + "."
+            + getPagedResponseTypeInnerName(method, typeTable, resourceField));
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/java/JavaSurfaceNamer.java]: I would prefer getAndSaveNicknameForStaticInnerClass to take two parameters:

1. Container type
2. Inner type

and rename it to getAndSaveNicknameForInnerType.

File: src/main/java/com/google/api/codegen/util/java/JavaTypeTable.java [hunk 20]
```diff
@@ -116,18 +136,29 @@ public String getAndSaveNicknameFor(TypeAlias alias) {
...
       // Short name clashes, use long name.
       return alias.getFullName();
     }
-    imports.put(alias.getFullName(), alias.getNickname());
+
+    if (alias.getParentFullName() == null) {
+      imports.put(alias.getFullName(), alias.getNickname());
+    } else {
+      staticImports.put(alias.getFullName(), alias.getNickname());
+    }
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/util/java/JavaTypeTable.java]: I think I would prefer to put more structure around the imports instead of creating two import lists. The two import lists is forcing a bunch of duplicated logic. I think it would also be nice to not force every language to support the notion of static imports - it might only be a Java thing.

File: src/main/java/com/google/api/codegen/transformer/java/JavaModelTypeNameConverter.java [hunk 21]
```diff
@@ -91,6 +91,11 @@ public JavaModelTypeNameConverter(String implicitPackageName) {
   }
 
   @Override
+  public TypeName getTypeName(String shortName) {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/java/JavaModelTypeNameConverter.java]: I think the shortness of this name is misleading as to what it does. By default, any methods in this set of classes that take a String treat it as a full name.

File: src/main/resources/com/google/api/codegen/java/common.snip (hunk 16 above)
Comment (michaelbausor) [inline src/main/resources/com/google/api/codegen/java/common.snip]: Removed

File: src/main/java/com/google/api/codegen/util/java/JavaTypeTable.java (hunk 17 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/util/java/JavaTypeTable.java]: Renamed

File: src/main/java/com/google/api/codegen/util/java/JavaTypeTable.java (hunk 20 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/util/java/JavaTypeTable.java]: Restructured to use TypeAlias instead of string, removed staticImports

File: src/main/java/com/google/api/codegen/util/go/GoTypeTable.java (hunk 18 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/util/go/GoTypeTable.java]: Go doesn't have an implicit package (at least, the GoTypeTable doesn't)

File: src/main/java/com/google/api/codegen/transformer/java/JavaSurfaceNamer.java (hunk 19 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/java/JavaSurfaceNamer.java]: Done

File: src/main/java/com/google/api/codegen/transformer/java/JavaModelTypeNameConverter.java (hunk 21 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/java/JavaModelTypeNameConverter.java]: Renamed

File: src/main/java/com/google/api/codegen/util/DynamicLangTypeTable.java [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  @Override
+  public String getAndSaveNicknameForInnerType(
+      String containerFullName, String innerTypeShortName) {
+    throw new UnsupportedOperationException("getAndSaveNicknameForStaticInnerClass not supported");
+  }
+
+  /**
+   * A set of PHP keywords and built-ins. keywords: http://php.net/manual/en/reserved.keywords.php
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/util/DynamicLangTypeTable.java]: Why php keywords here

**Comment by shinfan** [comment]
I am not sure if it is just me...but whenever a PR is large (more than 1000 LOC) I find the review is much harder to follow... So if possible, I would like to encourage to make progressive smaller PRs in the future.

**Comment by michaelbausor** [comment]
I agree about size - I will try to split changes like this up in future.

**Comment by michaelbausor** [comment]
PTAL - Passing tests when regenerating into api-client-staging

**Comment by shinfan** [comment]
Looks like the baseline tests are failing. Otherwise looks good.

**Comment by michaelbausor** [comment]
Fixed baselines.

File: src/main/java/com/google/api/codegen/transformer/ModelTypeFormatter.java [hunk 23]
```diff
@@ -26,6 +27,11 @@
  */
 public interface ModelTypeFormatter {
   /**
+   * Get the full name for the given short name, using the default package.
+   */
+  String getFullNameFor(String shortName);
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/ModelTypeFormatter.java]: I think this function should be more specific like the related ones: getImplicitPackageFullNameFor

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java [hunk 24]
```diff
@@ -163,6 +173,96 @@ private StaticLangXApiView generateXApi(SurfaceTransformerContext context) {
...
+    String name = namer.getPagedResponseWrappersClassName();
+    pagedResponseWrappers.name(name);
+
+    List<StaticLangPagedResponseView> pagedResponseWrappersList = new ArrayList<>();
+    for (Interface service : new InterfaceView().getElementIterable(model)) {
+      SurfaceTransformerContext context =
+          SurfaceTransformerContext.create(
+              service,
+              apiConfig,
+              createTypeTable(apiConfig.getPackageName()),
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java]: Why are you creating a new type table for each one?

File: src/test/java/com/google/api/codegen/testdata/java_main_library.baseline [hunk 25]
```diff
@@ -621,7 +622,7 @@ public class LibraryServiceApi implements AutoCloseable {
    * @param name The name of the shelf to delete.
    * @throws com.google.api.gax.grpc.ApiException if the remote call fails
    */
-  public final void deleteShelf(ShelfName name) {
+  public final void deleteShelf(String name) {
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/java_main_library.baseline]: Where did the ShelfName type go?

File: src/main/java/com/google/api/codegen/transformer/ImportTypeTransformer.java [hunk 26]
```diff
@@ -14,22 +14,28 @@
...
+  public List<ImportTypeView> generateImports(Map<String, TypeAlias> imports) {
     List<ImportTypeView> generatedImports = new ArrayList<>();
     for (String key : imports.keySet()) {
-      generatedImports.add(
-          ImportTypeView.newBuilder().fullName(key).nickname(imports.get(key)).build());
+      ImportTypeView.Builder importTypeView =
+          ImportTypeView.newBuilder().fullName(key).nickname(imports.get(key).getNickname());
+      if (imports.get(key).getParentFullName() != null) {
+        importTypeView.type(ImportType.StaticImport);
+      }
```
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/ImportTypeTransformer.java]: Done

File: src/main/java/com/google/api/codegen/util/DynamicLangTypeTable.java [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+  }
+
+  public boolean hasImports() {
+    return !getImports().isEmpty();
+  }
+
+  @Override
+  public String getAndSaveNicknameForInnerType(
+      String containerFullName, String innerTypeShortName) {
+    throw new UnsupportedOperationException("getAndSaveNicknameForStaticInnerClass not supported");
```
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/util/DynamicLangTypeTable.java]: Done

File: src/main/java/com/google/api/codegen/util/nodejs/NodeJSTypeTable.java [hunk 28]
```diff
@@ -14,94 +14,33 @@
...
- */
-public class NodeJSTypeTable implements TypeTable {
-  /**
-   * A bi-map from full names to short names indicating the import map.
-   */
-  private final BiMap<String, String> imports = HashBiMap.create();
-
-  private final String implicitPackageName;
+/** The TypeTable for NodeJS. */
+public class NodeJSTypeTable extends DynamicLangTypeTable {
```
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/util/nodejs/NodeJSTypeTable.java]: Done

File: src/main/java/com/google/api/codegen/transformer/ModelTypeFormatter.java (hunk 23 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/ModelTypeFormatter.java]: Done

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java (hunk 24 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java]: Should not be - fixed

File: src/test/java/com/google/api/codegen/testdata/java_main_library.baseline (hunk 25 above)
Comment (michaelbausor) [inline src/test/java/com/google/api/codegen/testdata/java_main_library.baseline]: Oops, this is a mistake - fixed

**Comment by michaelbausor** [comment]
Addressed comments and added a test for the pagedlistresponse iterateAllAs<X> method

File: src/main/java/com/google/api/codegen/util/nodejs/NodeJSTypeTable.java [hunk 29]
```diff
@@ -16,31 +16,76 @@
...
 import com.google.api.codegen.util.NamePath;
+import com.google.api.codegen.util.TypeAlias;
+import com.google.api.codegen.util.TypeName;
 import com.google.api.codegen.util.TypeTable;
 import com.google.common.collect.ImmutableSet;
+import java.util.Map;
 
 /** The TypeTable for NodeJS. */
-public class NodeJSTypeTable extends DynamicLangTypeTable {
+public class NodeJSTypeTable implements TypeTable {
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/util/nodejs/NodeJSTypeTable.java]: Why did you change this? This results a bunch of new overrides methods in this class.

File: src/main/resources/com/google/api/codegen/java/page_streaming_response.snip [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+        {@iterateMethods(pagedResponseClass)}
+      }
+      
+    @end
+  }
+@end
+
+@private iterateMethods(pagedResponseClass)
+  @join iterateMethod : pagedResponseClass.iterateMethods
+    public Iterable<{@iterateMethod.overloadResourceTypeName}> iterateAllAs{@iterateMethod.overloadResourceTypeName}() {
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/page_streaming_response.snip]: you are composing a type name in the template here (iterateAllAs*)

File: src/test/java/com/google/api/codegen/testdata/java_main_library.baseline [hunk 31]
```diff
@@ -1780,7 +1783,7 @@ public class LibraryServiceApi implements AutoCloseable {
    *   BookName namesElement = BookName.create("[SHELF_ID]", "[BOOK_ID]");
    *   List&lt;BookName&gt; names = Arrays.asList(namesElement);
    *   List&lt;ShelfName&gt; shelves = new ArrayList&lt;&gt;();
-   *   for (BookName element : libraryServiceApi.findRelatedBooks(names, shelves).iterateAllElements()) {
+   *   for (String element : libraryServiceApi.findRelatedBooks(names, shelves).iterateAllElements()) {
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/java_main_library.baseline]: I think the example loop should prefer the typed resource name instead of string

File: src/main/java/com/google/api/codegen/util/nodejs/NodeJSTypeTable.java [hunk 32]
```diff
@@ -14,45 +14,36 @@
...
- */
+/** The TypeTable for NodeJS. */
 public class NodeJSTypeTable implements TypeTable {
-  /**
-   * A bi-map from full names to short names indicating the import map.
-   */
-  private final BiMap<String, String> imports = HashBiMap.create();
 
-  private final String implicitPackageName;
+  private final DynamicLangTypeTable dynamicTypeTable;
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/util/nodejs/NodeJSTypeTable.java]: I won't block the review on this issue, but since the dynamic language type tables have more in common than they are different, I would have expected the composition to be arranged differently, where DynamicLangTypeTable implemented the TypeTable interface, and then had a strategy object with language-specific behavior.

File: src/main/java/com/google/api/codegen/util/nodejs/NodeJSTypeTable.java (hunk 29 above)
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/util/nodejs/NodeJSTypeTable.java]: @shinfan , it's better to look at the diff on the overall PR instead of this interdiff.

File: src/main/resources/com/google/api/codegen/java/page_streaming_response.snip (hunk 30 above)
Comment (michaelbausor) [inline src/main/resources/com/google/api/codegen/java/page_streaming_response.snip]: Done

File: src/test/java/com/google/api/codegen/testdata/java_main_library.baseline (hunk 31 above)
Comment (michaelbausor) [inline src/test/java/com/google/api/codegen/testdata/java_main_library.baseline]: Done

**Comment by michaelbausor** [comment]
Addressed PR comments, PTAL.

**Comment by garrettjonesgoogle** [comment]
LGTM; Since @shinfan has been giving comments too, he should probably also give an LGTM.

**Comment by michaelbausor** [comment]
Resolved merge conflicts, PTAL

**Comment by shinfan** [comment]
LGTM
